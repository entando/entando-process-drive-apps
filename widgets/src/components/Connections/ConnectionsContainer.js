/* eslint-disable no-console */
/* eslint-disable react/jsx-wrap-multilines */
import { Container, IconButton, Grid, Typography } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/core/styles';
import withStyles from '@material-ui/core/styles/withStyles';
import AddIcon from '@material-ui/icons/Add';
import React from 'react';
import PropTypes from 'prop-types';
import i18next from 'i18next';

import { DOMAINS, LOCAL } from 'api/constants';
import {
  getConnections,
  deleteConnection,
  testConnection,
  saveConnection,
  createConnection,
} from 'api/pda/connections';
import theme from 'theme';
import WidgetBox from 'components/common/WidgetBox';
import ConnectionItem from 'components/Connections/ConnectionItem';
import ConnectionItemSkeleton from 'components/Connections/ConnectionItemSkeleton';
import ConnectionForm from 'components/Connections/ConnectionForm';
import Notification from 'components/common/Notification';
import Loader from 'components/common/Loader';

const styles = {
  gridContainer: {
    marginTop: 10,
  },
};

const initialForm = {
  name: '',
  engine: '',
  schema: '',
  host: '',
  app: '',
  username: '',
  password: '',
};

class ConnectionsContainer extends React.Component {
  state = {
    connectionsList: [],
    showForm: false,
    formData: initialForm,
    listLoader: true,
    loader: false,
    notification: {
      message: '',
      type: 'error',
    },
  };

  componentDidMount = () => {
    const { serviceUrl } = this.props;

    if (!LOCAL) {
      // set the PDA domain to the URL passed via props
      DOMAINS.PDA = serviceUrl;
    }

    this.fetchConnections();
  };

  fetchConnections = async () => {
    const { payload: connectionsList } = await getConnections();

    this.setState({ connectionsList, listLoader: false });
  };

  handleAddButton = () => {
    this.setState({ showForm: true });
  };

  handleFormChange = form => ({ target: { value } }) => {
    const { formData: old } = this.state;
    const formData = { ...old, [form]: value };
    this.setState({ formData });
  };

  handleCancelForm = () => {
    this.setState({ formData: initialForm, showForm: false });
  };

  handleSave = async () => {
    const { formData } = this.state;
    try {
      const response = formData.edit
        ? await saveConnection(formData)
        : await createConnection(formData);

      if (response.payload) {
        this.setState(
          {
            showForm: false,
            notification: { message: i18next.t('messages.success.saved'), type: 'success' },
          },
          this.fetchConnections
        );
      }
    } catch (error) {
      this.setState({ notification: { message: error.message, type: 'error' } });
    }
  };

  handleEdit = form => () => {
    const formData = { ...initialForm, ...form, edit: true };
    this.setState({ formData, showForm: true });
  };

  handleDelete = con => async () => {
    this.setState({ loader: true });
    try {
      await deleteConnection(con);
      this.fetchConnections();
    } catch (error) {
      this.setState({ notification: { message: error.message, type: 'error' } });
    } finally {
      this.setState({ loader: false });
    }
  };

  handleTestConnection = con => async () => {
    const { connectionsList: list } = this.state;
    let connectionsList;

    this.setState({ loader: true });

    try {
      await testConnection(con);
      connectionsList = list.map(item =>
        item.name === con ? { ...item, online: 'online' } : item
      );
    } catch (error) {
      connectionsList = list.map(item =>
        item.name === con ? { ...item, online: 'offline' } : item
      );
    }
    this.setState({ connectionsList, loader: false });
  };

  handleCloseNotifications = () => {
    this.setState({ notification: { message: '' } });
  };

  render() {
    const { classes } = this.props;
    const { connectionsList, showForm, formData, listLoader, loader, notification } = this.state;

    return (
      <ThemeProvider theme={theme}>
        <Container>
          <WidgetBox
            title="Connections"
            topRightComp={
              <IconButton color="primary" onClick={this.handleAddButton}>
                <AddIcon />
              </IconButton>
            }
            open={showForm}
            hasDivider
          >
            <ConnectionForm
              form={formData}
              onChange={this.handleFormChange}
              onCancel={this.handleCancelForm}
              onSave={this.handleSave}
            />
          </WidgetBox>

          <Grid className={classes.gridContainer} container spacing={2}>
            {listLoader &&
              Array(4)
                .fill('')
                // eslint-disable-next-line react/no-array-index-key
                .map((_, i) => <ConnectionItemSkeleton key={i} />)}
            {connectionsList ? (
              connectionsList.map(connection => (
                <ConnectionItem
                  key={connection.name}
                  connection={connection}
                  onEdit={this.handleEdit}
                  onDelete={this.handleDelete}
                  onTest={this.handleTestConnection}
                />
              ))
            ) : (
              <Typography>No connections found.</Typography>
            )}
          </Grid>
          <Loader loading={loader} />
          <Notification
            message={notification.message}
            type={notification.type}
            onClose={this.handleCloseNotifications}
          />
        </Container>
      </ThemeProvider>
    );
  }
}

ConnectionsContainer.propTypes = {
  classes: PropTypes.shape({
    gridContainer: PropTypes.string,
  }).isRequired,
  serviceUrl: PropTypes.string,
};

ConnectionsContainer.defaultProps = {
  serviceUrl: '',
};

export default withStyles(styles)(ConnectionsContainer);
