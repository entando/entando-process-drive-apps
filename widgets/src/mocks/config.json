{
  "payload": {
    "code": "pda_task_list",
    "config": {
      "groups": "[{\"label\":\"Administrators\",\"key\":\"Administrators\",\"checked\":true},{\"label\":\"broker\",\"key\":\"broker\",\"checked\":true},{\"label\":\"admin\",\"key\":\"admin\",\"checked\":false},{\"label\":\"rest-all\",\"key\":\"rest-all\",\"checked\":false},{\"label\":\"kiemgmt\",\"key\":\"kiemgmt\",\"checked\":true},{\"label\":\"kie-server\",\"key\":\"kie-server\",\"checked\":true},{\"label\":\"appraiser\",\"key\":\"appraiser\",\"checked\":false},{\"label\":\"manager\",\"key\":\"manager\",\"checked\":true},{\"label\":\"supplier\",\"key\":\"supplier\",\"checked\":true},{\"label\":\"IT\",\"key\":\"IT\",\"checked\":true},{\"label\":\"PM\",\"key\":\"PM\",\"checked\":true}]",
      "options": "[{\"key\":\"newPageOnClick\",\"label\":\"Open new page on table row click\",\"checked\":false},{\"key\":\"showClaim\",\"label\":\"Show Claim Button\",\"checked\":true},{\"key\":\"showComplete\",\"label\":\"Show Complete Button\",\"checked\":true}]",
      "columns": "[{\"name\":\"task-is-skipable\",\"position\":0,\"isVisible\":false},{\"name\":\"task-parent-id\",\"position\":1,\"isVisible\":false},{\"name\":\"task-created-by\",\"position\":2,\"isVisible\":true},{\"name\":\"task-input-data.application.property.saleprice\",\"position\":3,\"isVisible\":true},{\"name\":\"task-input-data.TaskName\",\"position\":4,\"isVisible\":true},{\"name\":\"initiator\",\"position\":5,\"isVisible\":true},{\"name\":\"task-created-on\",\"position\":6,\"isVisible\":true},{\"name\":\"task-priority\",\"position\":7,\"isVisible\":false},{\"name\":\"task-proc-def-id\",\"position\":8,\"isVisible\":true},{\"name\":\"task-form\",\"position\":9,\"isVisible\":true},{\"name\":\"task-skippable\",\"position\":10,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.name\",\"position\":11,\"isVisible\":true},{\"name\":\"task-actual-owner\",\"position\":12,\"isVisible\":true},{\"name\":\"task-activation-time\",\"position\":13,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.address\",\"position\":14,\"isVisible\":true},{\"name\":\"task-input-data.application.errors\",\"position\":15,\"isVisible\":true},{\"name\":\"task-name\",\"position\":16,\"isVisible\":true},{\"name\":\"task-input-data.application.property.address\",\"position\":17,\"isVisible\":true},{\"name\":\"task-subject\",\"position\":18,\"isVisible\":true},{\"name\":\"containerId\",\"position\":19,\"isVisible\":true},{\"name\":\"task-container-id\",\"position\":20,\"isVisible\":true},{\"name\":\"processDefinitionId\",\"position\":21,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.ssn\",\"position\":22,\"isVisible\":true},{\"name\":\"task-description\",\"position\":23,\"isVisible\":true},{\"name\":\"task-input-data.Skippable\",\"position\":24,\"isVisible\":true},{\"name\":\"task-workitem-id\",\"position\":25,\"isVisible\":true},{\"name\":\"task-expiration-time\",\"position\":26,\"isVisible\":true},{\"name\":\"task-input-data.application.property.age\",\"position\":27,\"isVisible\":true},{\"name\":\"task-status\",\"position\":28,\"isVisible\":true},{\"name\":\"task-input-data.application.mortgageamount\",\"position\":29,\"isVisible\":true},{\"name\":\"task-process-instance-id\",\"position\":30,\"isVisible\":true},{\"name\":\"task-input-data.ActorId\",\"position\":31,\"isVisible\":true},{\"name\":\"task-process-id\",\"position\":32,\"isVisible\":true},{\"name\":\"task-input-data.application.property.locale\",\"position\":33,\"isVisible\":true},{\"name\":\"task-input-data.application.amortization\",\"position\":34,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.creditrating\",\"position\":35,\"isVisible\":true},{\"name\":\"application\",\"position\":36,\"isVisible\":true},{\"name\":\"task-input-data.NodeName\",\"position\":37,\"isVisible\":true},{\"name\":\"task-proc-inst-id\",\"position\":38,\"isVisible\":true},{\"name\":\"task-input-data.application.downpayment\",\"position\":39,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.annualincome\",\"position\":40,\"isVisible\":true},{\"name\":\"task-id\",\"position\":41,\"isVisible\":true}]",
      "knowledgeSource": "kieStaging",
      "process": "itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
    }
  },
  "metaData": {
    "status": "draft"
  },
  "errors": []
}
