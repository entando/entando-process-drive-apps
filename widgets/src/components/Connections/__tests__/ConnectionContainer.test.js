import { render, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import CONNECTIONS from 'mocks/pda/connections';
import ConnectionsContainer from 'components/Connections/ConnectionsContainer';

describe('<ConnectionsContainer />', () => {
  it('renders snapshot correctly', async () => {
    fetch.once(JSON.stringify(CONNECTIONS));

    const { container } = render(<ConnectionsContainer />);

    await wait(() => expect(container).toMatchSnapshot());
  });
});
