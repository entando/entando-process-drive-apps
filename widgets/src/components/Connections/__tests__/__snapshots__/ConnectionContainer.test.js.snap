// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectionsContainer /> renders snapshot correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root ControlledWidgetBox-widgetBox-10 MuiPaper-outlined"
    >
      <div
        class="MuiBox-root MuiBox-root-40"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2"
        >
          Connections
        </h2>
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Create new connection
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root ConnectionsContainer-gridContainer-1 MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <div
          class="MuiPaper-root MuiCard-root MuiPaper-outlined"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-avatar"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                k
              </div>
            </div>
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
              >
                kieStaging
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                pam
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root ConnectionItem-cardContent-244"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              style="padding-bottom: 10px;"
            >
              http://rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run:80/services/rest/
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
            >
              <b>
                Username: 
              </b>
               
              pamAdmin
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
            >
              <b>
                Timeout: 
              </b>
               
              60000
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Test connection
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              style="margin-left: auto;"
              tabindex="0"
              title="common.delete"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              title="common.edit"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <div
          class="MuiPaper-root MuiCard-root MuiPaper-outlined"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-avatar"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                k
              </div>
            </div>
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
              >
                kie2Server
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                pam
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root ConnectionItem-cardContent-244"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              style="padding-bottom: 10px;"
            >
              http://rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run:80/services/rest/
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
            >
              <b>
                Username: 
              </b>
               
              pamAdmin
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
            >
              <b>
                Timeout: 
              </b>
               
              60000
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Test connection
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              style="margin-left: auto;"
              tabindex="0"
              title="common.delete"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              title="common.edit"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <div
          class="MuiPaper-root MuiCard-root MuiPaper-outlined"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-avatar"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                t
              </div>
            </div>
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
              >
                test
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                pam
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root ConnectionItem-cardContent-244"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              style="padding-bottom: 10px;"
            >
              http://rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run:80/services/rest/
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
            >
              <b>
                Username: 
              </b>
               
              pamAdmin
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
            >
              <b>
                Timeout: 
              </b>
               
              60000
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Test connection
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              style="margin-left: auto;"
              tabindex="0"
              title="common.delete"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              title="common.edit"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
